src/spect_experimental/utilities/ctac_to_mu_values.cxx

```diff
@@ -0,0 +1,224 @@
+  \endcode
+*/
+
+#include <algorithm>
+#include <cctype>
+
+#include "stir/IO/ITKImageInputFileFormat.h"
+#include "stir/info.h"
+#include "stir/Succeeded.h"
+#include "stir/IO/OutputFileFormat.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/DiscretisedDensity.h"
+#include "stir/modelling/ParametricDiscretisedDensity.h"
+#include "stir/DynamicDiscretisedDensity.h"
+#include "stir/VoxelsOnCartesianGrid.h"
+#include "stir/stream.h"
+#include "stir/getopt.h"
+
+
+#include <nlohmann/json.hpp>
+
+USING_NAMESPACE_STIR
+
+typedef DiscretisedDensity<3,float> FloatImageType;
+
+Succeeded apply_bilinear_scaling_to_HU(
+    const std::unique_ptr<FloatImageType> &input_image_sptr,
+    const nlohmann::json &transform,
+    std::shared_ptr<FloatImageType> output_image_sptr){
+
+  FloatImageType::full_iterator out_iter = output_image_sptr->begin_all();
+  FloatImageType::const_full_iterator in_iter = input_image_sptr->begin_all_const();
+
+  const float a1 = transform["a1"];
+  const float b1 = transform["b1"];
+
+  const float a2 = transform["a2"];
+  const float b2 = transform["b2"];
+
+  //std::cout << transform.dump(4);
+
+  while( in_iter != input_image_sptr->end_all_const())
+  {
+    if (*in_iter<0.f) {
+      float mu = a1 + b1 *(*in_iter);
+      *out_iter = (mu < 0.0f) ? 0.0f : mu;
+    } else {
+      *out_iter = a2 + b2 * (*in_iter);
+    }
+
+    ++in_iter; ++out_iter;
+  }
+
+  return Succeeded::yes;
+}
+
+
+int main(int argc, char * argv[])
+{
+  USING_NAMESPACE_STIR;
+  const char * output_filename = 0;
+  const char * input_filename = 0;
+  const char * slope_filename = 0;
+  const char * manufacturer_name = 0;
+  const char * keV_str = 0;
+
+  const char * const usage = "ctac_to_mu_values -o output_filename -i input_dicom_slice -j slope_filename -m manufacturer_name -k target_energy\n";
+  opterr = 0;
+  {
+    char c;
+
+    while ((c = getopt (argc, argv, "i:o:j::m:k:?")) != -1)
+      switch (c)
+	{
+	case 'i':
+	  input_filename = optarg;
+	  break;
+	case 'o':
+	  output_filename = optarg;
+	  break;
+	case 'j':
+      slope_filename = optarg;
+	  break;
+    case 'm':
+      manufacturer_name = optarg;
+      break;
+	case 'k':
+      keV_str = optarg;
+      break;
+	case '?':
+	  std::cerr << usage;
+	  return EXIT_FAILURE;
+	default:
+	  if (isprint (optopt))
+	    fprintf (stderr, "Unknown option `-%c'.\n", optopt);
+	  else
+	    fprintf (stderr,
+		     "Unknown option character `\\x%x'.\n",
+		     optopt);
+	  std::cerr << usage;
+	  return EXIT_FAILURE;
+	}
+  }
+
+  if (output_filename==0 || input_filename==0 || slope_filename==0 || manufacturer_name==0 || keV_str==0 )
+    {
+      std::cerr << usage;
+      return EXIT_FAILURE;
+    }
+
+  //Read slope file
+  std::ifstream slope_json_file_stream(slope_filename);
+  nlohmann::json slope_json;
+  slope_json_file_stream >> slope_json;
+
+  //Put user-specified manufacturer into upper case.
+  std::string manufacturer = manufacturer_name;
+  std::locale loc;
```

Comment:
could you move L145 to L179 to a separate function? Will make it easier for Python etc later on.

Reply:
Done aa2525a.